### ISSUE TITLE
[DOCS][MANIFEST] Revise manifest.md into complete agent-ready design spec

### ISSUE THREAD

**Issue by lcy-seso** [tile-ai/TileOPs#743]
Parent: #741

### Symptom / Motivation

`docs/manifest.md` has accumulated inconsistencies between spec and implementation, missing rules, and underspecified semantics. It cannot reliably guide an agent to produce correct manifest entries or generate code from the manifest.

Seven specific gaps were identified through discussion of #743 (original) and #744:

1. `params` format: spec says list, implementation uses dict — ambiguous positional semantics
2. No `status` field — cannot express spec-first ops without implementation
3. Roofline variables (`M`, `N`, `C`, `L`, `elem_bytes`) used but never declared — agent cannot determine valid bindings
4. `keepdim` missing from reduction op examples in #744, contradicting manifest.md's own Reduce example
5. No rule enforcing PyTorch interface alignment — risk of inventing non-standard parameters
6. Scan ops (cumsum, cumprod) conflated with reductions despite different shape semantics
7. `dtype_combos` defined but usage rules unclear — validator cannot determine when Cartesian product vs explicit enumeration applies

### Root Cause Analysis

The manifest spec was written incrementally alongside early op implementations. Rules were added per-family without cross-cutting consistency review. The spec was never validated as "sufficient for an agent to work from alone."

### Related Files

- `docs/manifest.md`
- `tileops/ops_manifest.yaml`

## Goal

Revise `docs/manifest.md` into a complete, unambiguous spec that an agent can follow to: (1) write correct manifest entries, (2) generate `Op.forward()` / `Op.infer_shape()`, (3) produce validators and test generators.

### Action 1: R1 — Ordered dict with explicit ordering contract

Modify R1 from "Dict, not list" to:

> **R1. Ordered dict.** `inputs`, `outputs`, and `params` are dicts keyed by name. Key order is significant — it defines the function signature position.
>
> **Implementation constraint:** YAML spec does not guarantee mapping order. This project relies on Python 3.7+ `dict` insertion-order preservation via PyYAML `yaml.safe_load()`. All manifest consumers (validators, codegen, agents) MUST use an order-preserving parser.
>
> **Agent rule:** When generating `Op.forward()`, `Op.infer_shape()`, or test code from the manifest, the agent MUST preserve the key order as the positional argument order. Reordering is a breaking change.

Changes:
- Update R1 rule text
- Update Signature schema comment: `params` from `# list` to `# ordered dict — param name → {type, default?}`
- Convert all residual list-format params examples in manifest.md to dict format
- `ops_manifest.yaml` — zero changes (already uses dict)

### Action 2: Add `status` field with gating rule

Add to Entry Structure table:

| Field | Required | Description |
|---|---|---|
| `status` | yes | `spec` or `implemented` |

Add rule:

> **R13. Status gating.** `status` is required on every entry. When `status: spec`, `source` paths declare intended locations — file-existence validation is skipped. When `status: implemented`, all `source` paths must resolve to existing files.

No default value — every entry must declare explicitly.

### Action 3: Roofline `vars` binding mechanism

Add rule:

> **R14. Roofline variable binding.**
>
> **Built-in variables:** `elem_bytes` — byte size of the primary input's dtype. Always available.
>
> **Auto-bound from `shape`:** When a tensor declares `shape: "[M, K]"`, dimension names are directly available as roofline variables. `vars` may be omitted.
>
> **Explicit `vars`:** When tensors have arbitrary rank (no `shape` field), roofline must include a `vars` mapping. Each entry is `name: "Python expression"`.
>
> **`vars` evaluation context:** tensor shapes (`x.shape`, `x.shape[i]`, `x.shape[start:end]`, `x.ndim`), params by name (`dim`, `groups`), helpers (`product(iterable)`), arithmetic (`+`, `-`, `*`, `//`, `**`, `%`), built-ins (`range()`, `len()`, list comprehensions).
>
> **`flops`/`bytes` evaluation context (unchanged):** Numeric constants, bound variables, arithmetic, `log2`, `ceil`, `floor`. No shape access — all shape-dependent values must come through `vars`.
>
> **Mixed dtype:** `elem_bytes` binds to the primary input. Other dtypes use literal byte counts (e.g., `4` for float32).

Add four case examples in the Roofline section:

```yaml
# Case 1: Fixed rank — vars omitted, shape names auto-bind
roofline:
  flops: "2 * M * N * K"
  bytes: "(M * K + K * N + M * N) * elem_bytes"

# Case 2: Arbitrary rank, single-dim reduction
roofline:
  vars:
    M: "product(x.shape[:dim])"
    N: "x.shape[dim]"
  flops: "4 * M * N"
  bytes: "(2 * M * N + N) * elem_bytes"

# Case 3: Spatial norm
roofline:
  vars:
    C: "x.shape[1]"
    L: "product(x.shape) // x.shape[1]"
  flops: "10 * C * L"
  bytes: "2 * C * L * elem_bytes + 4 * C * 4"

# Case 4: Multi-dim reduction
roofline:
  vars:
    M: "product(x.shape[i] for i in range(x.ndim) if i not in dim)"
    N: "product(x.shape[i] for i in dim)"
  flops: "M * N"
  bytes: "(M * N + M) * elem_bytes"
```

Update Roofline table to include `vars` mode:

| Mode | Format | When |
|---|---|---|
| Inline (no vars) | `flops`/`bytes` only | Fixed-rank ops with `shape` dimension names |
| Inline + vars | `vars` + `flops`/`bytes` | Arbitrary-rank ops needing dimension binding |
| Func | `func: "module.path"` | Tier 2: complex formulas needing Python code |

### Action 4: Reduction ops must include `keepdim`

All reduction ops (sum, mean, amax, amin, prod, argmax, argmin, all, any, count_nonzero, l1_norm, l2_norm, inf_norm, var, std, var_mean, softmax, log_softmax, logsumexp) must include:

```yaml
params:
  dim: {type: "int | list[int]"}
  keepdim: {type: bool, default: false}
shape_rules:
  - "y.ndim == x.ndim if keepdim else x.ndim - len(dim)"
  - "y.shape[i] == 1 if i in dim and keepdim else x.shape[i]"
```

Exception: softmax/log_softmax/logsumexp where output shape always equals input shape — `keepdim` is not applicable.

Note: manifest.md's existing Reduce example is already correct. #744's examples need revision.

### Action 5: R15 — PyTorch interface alignment

Add rule:

> **R15. PyTorch interface alignment.** Op signatures must match PyTorch's public API conventions (parameter names, parameter set, semantics). Do not invent parameters that PyTorch does not expose. When PyTorch has no direct equivalent, follow the closest community convention (numpy, scipy).

Consequence: spatial-norm ops (batchnorm, groupnorm, instancenorm) do NOT get a `dim` parameter — matching PyTorch's `nn.BatchNorm2d`, `nn.GroupNorm`, `nn.InstanceNorm2d`.

### Action 6: Scan ops use `family: scan`

`cumsum_fwd` and `cumprod_fwd` use `family: scan`, not `family: reduction`. Reduce and scan have different semantics in functional programming: reduce collapses a dimension, scan preserves shape with prefix accumulation.

Update Tier Model table to include `scan` in families list.

### Action 7: Clarify `dtype_combos` semantics (R4)

Revise R4 to:

> **R4. `dtype_combos`.** Enumerates the actually supported cross-tensor dtype combinations.
>
> - **When present:** Exhaustive source of truth. Only listed combinations are valid. Validator and test generator use this list, not the Cartesian product.
> - **When absent:** All combinations from the Cartesian product of per-tensor `dtype` fields are assumed supported. Validator must verify all of them.
>
> Use `dtype_combos` when the supported set is a strict subset of the Cartesian product (e.g., mixed-precision GEMM supports 3 of 4 possible combinations). Omit when all combinations are valid (e.g., all inputs are `same_as(x)`).

## Constraints

- Changes are limited to `docs/manifest.md` only — no implementation code changes
- Existing `ops_manifest.yaml` entries must remain valid under the revised spec
- Rules must be numbered sequentially (R1–R15) without gaps

## Acceptance Criteria

- [ ] AC-1: R1 updated to ordered-dict with implementation constraint and agent rule documented
- [ ] AC-2: `status` field added to Entry Structure with R13 gating rule
- [ ] AC-3: R14 roofline `vars` mechanism with evaluation context and 4 case examples
- [ ] AC-4: Reduce example in manifest.md includes `keepdim`; softmax exception noted
- [ ] AC-5: R15 PyTorch interface alignment rule added
- [ ] AC-6: `scan` family added to Tier Model table
- [ ] AC-7: R4 revised with Cartesian-product-vs-explicit semantics
- [ ] AC-8: All examples in manifest.md use dict format for params (no residual list format)
- [ ] AC-9: Modified files pass existing tests
